File: Python/_0560_Subarray_Sum_Equals_K/Solution.py
# coding=utf-8
# Date: 2021/5/29 14:28
from typing import List


# 执行用时：112 ms, 在所有 Python3 提交中击败了22.72%的用户
# 内存消耗：17.1 MB, 在所有 Python3 提交中击败了17.98%的用户
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        n = len(nums)
        s = [0 for _ in range(n + 1)]
        for i in range(1, n + 1):
            s[i] = s[i - 1] + nums[i - 1]
        hash = {}
        hash[0] = 1
        res = 0
        for i in range(1, n + 1):
            if s[i] - k in hash:
                res += hash[s[i] - k]
            if s[i] in hash:
                hash[s[i]] += 1
            else:
                hash[s[i]] = 1
        return res


if __name__ == "__main__":
    print(Solution().subarraySum([1, 1, 1], 2))  # 2

File: Java/src/_0560_Subarray_Sum_Equals_K/Solution.java
package _0560_Subarray_Sum_Equals_K;

import java.util.HashMap;

/**
 * Date: 2021/5/29 14:24
 * 执行用时：19 ms, 在所有 Java 提交中击败了99.62%的用户
 * 内存消耗：41.4 MB, 在所有 Java 提交中击败了29.58%的用户
 */
public class Solution {

    public int subarraySum(int[] nums, int k) {
        int n = nums.length;
        int[] s = new int[n + 1];
        for (int i = 1; i <= n; i++) s[i] = s[i - 1] + nums[i - 1];
        HashMap<Integer, Integer> hash = new HashMap<>();
        hash.put(0, 1);
        int res = 0;
        for (int i = 1; i <= n; i++) {
            res += hash.getOrDefault(s[i] - k, 0);
            hash.put(s[i], hash.getOrDefault(s[i], 0) + 1);
        }
        return res;
    }

    public static void main(String[] args) {

        System.out.println((new Solution()).subarraySum(new int[]{1, 1, 1}, 2));  // 2
    }
}
